File: packages/workflow/src/review.ts
import type { OrderLine } from "./orders.js";

export interface InventorySnapshot {
  matchKey: string;
  wdtSpecNo: string;
  mainAvailableStock: number;
  nearExpiryAvailableStock: number;
  defectAvailableStock?: number;
  otherAvailableStock?: number;
  warehouseBreakdown?: string;
}

export type ReviewStatus = "库存充足" | "部分满足" | "库存不足" | "未匹配";

export interface ReviewLine {
  orderNoticeNo: string;
  excelRow: number;
  storeNo: string;
  storeName: string;
  uploadTime: string;
  orderApprovalNo: string;
  readingStatus: string;
  deliveryMode: string;
  orderStatus: string;
  deliveryTarget: string;
  category: string;
  orderDate: string;
  deadlineDate: string;
  salesperson: string;
  maker: string;
  madeAt: string;
  sourceReviewer: string;
  externalGoodsCode: string;
  externalBarcode: string;
  externalGoodsName: string;
  originalSpec: string;
  transportSpec: string;
  orderBoxQty: string;
  orderQty: number;
  taxExcludedUnitPrice: string;
  contractPrice: string;
  taxIncludedUnitPrice: string;
  discountRate: string;
  shelfLifeDays: string;
  receivedQty: string;
  giftRate: string;
  td: string;
  da: string;
  pd: string;
  spd: string;
  rebate: string;
  orderRawJson: string;
  wdtSpecNo: string;
  mainAvailableBefore: number;
  nearExpiryAvailableBefore: number;
  defectAvailableBefore: number;
  otherAvailableBefore: number;
  warehouseBreakdown: string;
  suggestedMainQty: number;
  suggestedNearExpiryQty: number;
  suggestedShipQty: number;
  remainingAfter: number;
  status: ReviewStatus;
}

export function buildReviewLines(orderLines: OrderLine[], inventoryByBarcode: Map<string, InventorySnapshot>): ReviewLine[] {
  const remainingMain = new Map<string, number>();
  const remainingNearExpiry = new Map<string, number>();
  for (const [key, snapshot] of inventoryByBarcode) {
    remainingMain.set(key, snapshot.mainAvailableStock);
    remainingNearExpiry.set(key, snapshot.nearExpiryAvailableStock);
  }

  return orderLines.map((line): ReviewLine => {
    const key = line.externalBarcode;
    const snapshot = inventoryByBarcode.get(key);
    if (!snapshot) {
      return {
        orderNoticeNo: line.orderNoticeNo,
        excelRow: line.excelRow,
        storeNo: line.storeNo,
        storeName: line.storeName,
        uploadTime: line.uploadTime,
        orderApprovalNo: line.orderApprovalNo,
        readingStatus: line.readingStatus,
        deliveryMode: line.deliveryMode,
        orderStatus: line.orderStatus,
        deliveryTarget: line.deliveryTarget,
        category: line.category,
        orderDate: line.orderDate,
        deadlineDate: line.deadlineDate,
        salesperson: line.salesperson,
        maker: line.maker,
        madeAt: line.madeAt,
        sourceReviewer: line.sourceReviewer,
        externalGoodsCode: line.externalGoodsCode,
        externalBarcode: line.externalBarcode,
        externalGoodsName: line.externalGoodsName,
        originalSpec: line.spec,
        transportSpec: line.transportSpec,
        orderBoxQty: line.orderBoxQty,
        orderQty: line.orderQty,
        taxExcludedUnitPrice: line.taxExcludedUnitPrice,
        contractPrice: line.contractPrice,
        taxIncludedUnitPrice: line.unitPriceTaxIncluded,
        discountRate: line.discountRate,
        shelfLifeDays: line.shelfLifeDays,
        receivedQty: line.receivedQty,
        giftRate: line.giftRate,
        td: line.td,
        da: line.da,
        pd: line.pd,
        spd: line.spd,
        rebate: line.rebate,
        orderRawJson: JSON.stringify(line.raw),
        wdtSpecNo: "",
        mainAvailableBefore: 0,
        nearExpiryAvailableBefore: 0,
        defectAvailableBefore: 0,
        otherAvailableBefore: 0,
        warehouseBreakdown: "",
        suggestedMainQty: 0,
        suggestedNearExpiryQty: 0,
        suggestedShipQty: 0,
        remainingAfter: 0,
        status: "未匹配",
      };
    }

    const mainBefore = remainingMain.get(key) ?? 0;
    const nearBefore = remainingNearExpiry.get(key) ?? 0;
    const defectBefore = snapshot.defectAvailableStock ?? 0;
    const otherBefore = snapshot.otherAvailableStock ?? 0;
    const suggestedMainQty = Math.min(line.orderQty, mainBefore);
    const suggestedNearExpiryQty = Math.min(line.orderQty - suggestedMainQty, nearBefore);
    const suggestedShipQty = suggestedMainQty + suggestedNearExpiryQty;
    const mainAfter = mainBefore - suggestedMainQty;
    const nearAfter = nearBefore - suggestedNearExpiryQty;
    remainingMain.set(key, mainAfter);
    remainingNearExpiry.set(key, nearAfter);

    let status: ReviewStatus = "库存不足";
    if (suggestedShipQty >= line.orderQty) status = "库存充足";
    else if (suggestedShipQty > 0) status = "部分满足";

    return {
      orderNoticeNo: line.orderNoticeNo,
      excelRow: line.excelRow,
      storeNo: line.storeNo,
      storeName: line.storeName,
      uploadTime: line.uploadTime,
      orderApprovalNo: line.orderApprovalNo,
      readingStatus: line.readingStatus,
      deliveryMode: line.deliveryMode,
      orderStatus: line.orderStatus,
      deliveryTarget: line.deliveryTarget,
      category: line.category,
      orderDate: line.orderDate,
      deadlineDate: line.deadlineDate,
      salesperson: line.salesperson,
      maker: line.maker,
      madeAt: line.madeAt,
      sourceReviewer: line.sourceReviewer,
      externalGoodsCode: line.externalGoodsCode,
      externalBarcode: line.externalBarcode,
      externalGoodsName: line.externalGoodsName,
      originalSpec: line.spec,
      transportSpec: line.transportSpec,
      orderBoxQty: line.orderBoxQty,
      orderQty: line.orderQty,
      taxExcludedUnitPrice: line.taxExcludedUnitPrice,
      contractPrice: line.contractPrice,
      taxIncludedUnitPrice: line.unitPriceTaxIncluded,
      discountRate: line.discountRate,
      shelfLifeDays: line.shelfLifeDays,
      receivedQty: line.receivedQty,
      giftRate: line.giftRate,
      td: line.td,
      da: line.da,
      pd: line.pd,
      spd: line.spd,
      rebate: line.rebate,
      orderRawJson: JSON.stringify(line.raw),
      wdtSpecNo: snapshot.wdtSpecNo,
      mainAvailableBefore: mainBefore,
      nearExpiryAvailableBefore: nearBefore,
      defectAvailableBefore: defectBefore,
      otherAvailableBefore: otherBefore,
      warehouseBreakdown: snapshot.warehouseBreakdown ?? "",
      suggestedMainQty,
      suggestedNearExpiryQty,
      suggestedShipQty,
      remainingAfter: mainAfter + nearAfter,
      status,
    };
  });
}

File: src/jy_trade/review.py
from __future__ import annotations

from dataclasses import asdict, dataclass

from .orders import OrderLine


@dataclass(frozen=True)
class InventorySnapshot:
    match_key: str
    wdt_spec_no: str
    main_available_stock: float = 0.0
    near_expiry_available_stock: float = 0.0


@dataclass(frozen=True)
class ReviewLine:
    order_notice_no: str
    excel_row: int
    store_no: str
    store_name: str
    upload_time: str
    external_barcode: str
    external_goods_name: str
    order_qty: float
    wdt_spec_no: str
    main_available_before: float
    near_expiry_available_before: float
    suggested_main_qty: float
    suggested_near_expiry_qty: float
    suggested_ship_qty: float
    remaining_after: float
    status: str

    def to_dict(self) -> dict[str, object]:
        return asdict(self)


def build_review_lines(
    order_lines: list[OrderLine],
    inventory_by_barcode: dict[str, InventorySnapshot],
) -> list[ReviewLine]:
    remaining_main = {
        key: snapshot.main_available_stock
        for key, snapshot in inventory_by_barcode.items()
    }
    remaining_near_expiry = {
        key: snapshot.near_expiry_available_stock
        for key, snapshot in inventory_by_barcode.items()
    }

    review_lines: list[ReviewLine] = []
    for line in order_lines:
        key = line.external_barcode
        snapshot = inventory_by_barcode.get(key)
        if snapshot is None:
            review_lines.append(
                ReviewLine(
                    order_notice_no=line.order_notice_no,
                    excel_row=line.excel_row,
                    store_no=line.store_no,
                    store_name=line.store_name,
                    upload_time=line.upload_time,
                    external_barcode=line.external_barcode,
                    external_goods_name=line.external_goods_name,
                    order_qty=line.order_qty,
                    wdt_spec_no="",
                    main_available_before=0.0,
                    near_expiry_available_before=0.0,
                    suggested_main_qty=0.0,
                    suggested_near_expiry_qty=0.0,
                    suggested_ship_qty=0.0,
                    remaining_after=0.0,
                    status="未匹配",
                )
            )
            continue

        main_before = remaining_main[key]
        near_before = remaining_near_expiry[key]
        main_qty = min(line.order_qty, main_before)
        shortage_after_main = line.order_qty - main_qty
        near_qty = min(shortage_after_main, near_before)
        suggested_ship_qty = main_qty + near_qty

        remaining_main[key] = main_before - main_qty
        remaining_near_expiry[key] = near_before - near_qty
        remaining_total = remaining_main[key] + remaining_near_expiry[key]

        if suggested_ship_qty >= line.order_qty:
            status = "库存充足"
        elif suggested_ship_qty > 0:
            status = "部分满足"
        else:
            status = "库存不足"

        review_lines.append(
            ReviewLine(
                order_notice_no=line.order_notice_no,
                excel_row=line.excel_row,
                store_no=line.store_no,
                store_name=line.store_name,
                upload_time=line.upload_time,
                external_barcode=line.external_barcode,
                external_goods_name=line.external_goods_name,
                order_qty=line.order_qty,
                wdt_spec_no=snapshot.wdt_spec_no,
                main_available_before=main_before,
                near_expiry_available_before=near_before,
                suggested_main_qty=main_qty,
                suggested_near_expiry_qty=near_qty,
                suggested_ship_qty=suggested_ship_qty,
                remaining_after=remaining_total,
                status=status,
            )
        )

    return review_lines
